### PR TITLE
remove workarounds from custom pod controller and refactor code

### PR DESCRIPTION
**Pull request by abhipth**
*Description of changes:*

- Removing workaround from the memory optimized Pod controller to pass Delete Event with request.
- Renaming file and structs, interfaces names to more descriptive names.
- Refactor code so it's easier to add new resource like IPv4 Address on top of Pod ENI.

**Testing for the Change**
- Ran Integration test in loop (x5 times) - Passed
- Ensure memory doesn't increase as the Number of Pods increases - Passed
- Ensure existing Pod's networking is not removed once the controller restarts - Passed

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/handler/handler.go [hunk 1]
```diff
@@ -24,7 +26,7 @@ import (
 // processed on demand would fit into this category. For instance, Branch ENIs are tied to the Security
 // Group required by the pod which we would know only after receiving the pod request.
 type Handler interface {
-	CanHandle(resourceName string) bool
-	HandleCreate(resourceName string, requestCount int, pod *v1.Pod) (ctrl.Result, error)
-	HandleDelete(resourceName string, pod *v1.Pod) (ctrl.Result, error)
+	GetProvider() provider.ResourceProvider
```
Comment (orsenthil) [inline pkg/handler/handler.go:29]: With the terminology and the design here, I have some questions.

- Usually a Handler exposes the methods for handling operations (like created, delete, update) here.
- Adding `GetProvider()` to Handler interface, seems to me to add two different functions in the interface.

I understood the behavior change, this is a common the code organization or interface design.

**Review by orsenthil — COMMENTED** [review]
I am not an authority on this code base. I looked it, trying to understand the change and adding comments.

Please take other reviewer's comments over mine.

File: pkg/handler/handler.go (hunk 1 above)
Comment (abhipth) [reply pkg/handler/handler.go:29]: Agreed, will remove the GetProvider() from the interface.

File: controllers/custom/builder.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	return b
+}
+
+func NewControllerManagedBy(mgr manager.Manager) *Builder {
+	return &Builder{mgr: mgr}
+}
+
+// Complete adds the controller to manager's Runnable. The Controller
+// runnable will start when the manager starts
+func (b *Builder) Complete(reconciler Reconciler) error {
```
Comment (orsenthil) [inline controllers/custom/builder.go:88]: `Complete` sems very generic a name.. How about something specific like `AddController` or anything which is specific.

File: pkg/k8s/wrapper.go [hunk 3]
```diff
@@ -75,20 +52,6 @@ var (
...
-	)
-
-	getPodFromAPIServeErrCount = prometheus.NewCounter(
-		prometheus.CounterOpts{
-			Name: "get_pod_from_api_server_err_count",
-			Help: "The number of requests that failed to get the pod directly from API Server",
-		},
-	)
-
 	broadcastPodEventCallCount = prometheus.NewCounter(
```
Comment (orsenthil) [inline pkg/k8s/wrapper.go:55]: This seems unused local variables. Will `go lint` catch these? Might be a good idea to run it find other unused local variables.

**Review by orsenthil — COMMENTED** [review]
* This https://github.com/aws/amazon-vpc-resource-controller-k8s/pull/44/files#r690825698 seems to be only significant comment that could change the code a little again.

* Rest of the comments are to improve readability and comprehensibility.

* There are unused local variables through out the code, perhaps linter will help remove them all.

**Comment by orsenthil** [comment]
@abhipth - I am done reviewing. Please address these comments and I could review again and approve it.

File: controllers/core/pod_controller_test.go [hunk 4]
```diff
@@ -16,19 +16,19 @@ package controllers
 import (
 	"testing"
 
-	mock_controller "github.com/aws/amazon-vpc-resource-controller-k8s/mocks/amazon-vcp-resource-controller-k8s/controllers/custom"
-	mock_handler "github.com/aws/amazon-vpc-resource-controller-k8s/mocks/amazon-vcp-resource-controller-k8s/pkg/handler"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/controllers/custom"
+	mockhandler "github.com/aws/amazon-vpc-resource-controller-k8s/mocks/amazon-vcp-resource-controller-k8s/pkg/handler"
```
[the remark this answers is not in the document]
Comment (abhipth) [reply controllers/core/pod_controller_test.go:20]: thanks for pointing this out, I have switched to the default mock naming convention (snake case).

File: controllers/custom/builder.go (hunk 2 above)
Comment (abhipth) [reply controllers/custom/builder.go:88]: I tried to keep it inline with upstream controller's interface -  https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.9/pkg/builder/controller.go#L167-L171

I am fine with renaming it to AddController too.

File: pkg/k8s/wrapper.go (hunk 3 above)
Comment (abhipth) [reply pkg/k8s/wrapper.go:55]: [golint](https://github.com/golang/lint) seems to have been deprecated, found one more tool [go-tools](https://github.com/dominikh/go-tools) which do the job.

File: controllers/custom/builder.go (hunk 2 above)
Comment (orsenthil) [reply controllers/custom/builder.go:88]: Consistency with upstream is better. So, +1 in keeping it so.
I didn't realize that both are enabling a builder pattern. I looked at other methods and then realized it.

**Comment by orsenthil** [comment]
The unit tests failure is due to this upstream issue - https://github.com/kubernetes-sigs/kubebuilder/issues/2311

File: controllers/core/pod_controller.go [hunk 5]
```diff
@@ -146,27 +125,23 @@ func getAggregateResources(pod *v1.Pod) map[string]int64 {
...
-
-// SetupWithManager sets controller to listen on the delete events
-func (d *DeleteReconciler) SetupWithManager(manager ctrl.Manager) error {
-	// Delete event controller
-	deleteController, err := controller.New("pod-delete", manager, controller.Options{Reconciler: d})
-	if err != nil {
-		return err
-	}
-	return deleteController.Watch(&source.Channel{Source: d.DeleteEventChannel}, &handler.EnqueueRequestForObject{})
+// SetupWithManager adds the custom Pod controller's runnable to the mangers
```
Comment (orsenthil) [inline controllers/core/pod_controller.go:128]: `to the managers list`

File: controllers/core/pod_controller_test.go [hunk 6]
```diff
@@ -16,19 +16,19 @@ package controllers
...
 	mock_node "github.com/aws/amazon-vpc-resource-controller-k8s/mocks/amazon-vcp-resource-controller-k8s/pkg/node"
-	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/handler"
+	mock_resource "github.com/aws/amazon-vpc-resource-controller-k8s/mocks/amazon-vcp-resource-controller-k8s/pkg/resource"
```
Comment (orsenthil) [inline controllers/core/pod_controller_test.go:22]: Still a minor inconsistency here. `mockhandler` vs `mock_resource`

File: pkg/resource/manager.go [hunk 7]
```diff
@@ -29,11 +29,13 @@ import (
...
 
 type Manager struct {
-	handlers map[string]handler.Handler
+	handlers  map[string]handler.Handler
+	providers []provider.ResourceProvider
 }
 
 type ResourceManager interface {
 	GetResourceHandlers() []handler.Handler
+	GetResourceProviders() []provider.ResourceProvider
```
Comment (orsenthil) [inline pkg/resource/manager.go:38]: +1. yeah, this is better.

**Review by orsenthil — APPROVED** [review]
LGTM. thank you.

File: controllers/core/pod_controller_test.go (hunk 6 above)
Comment (orsenthil) [reply controllers/core/pod_controller_test.go:22]: these comments could be addressed in next PR. This is approved and can be merged.

Comment (abhipth) [reply controllers/core/pod_controller_test.go:22]: Will approve in the next PR. Thanks!

File: controllers/core/pod_controller.go (hunk 5 above)
Comment (abhipth) [reply controllers/core/pod_controller.go:128]: Will update in next PR.

**Pull request merged**